[@typescript-eslint/require-await] false positive on async generators without `await`

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"

Please don't ignore this template.

If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
Are you opening an issue because the rule you're trying to use is not found?
🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨
1) Check the releases log: https://github.com/typescript-eslint/typescript-eslint/releases
    -  If the rule isn't listed there, then chances are it hasn't been released to the main npm tag yet.
2) Try installing the `canary` tag: `npm i @typescript-eslint/eslint-plugin@canary`.
    - The canary tag is built for every commit to master, so it contains the bleeding edge build.
3) If ESLint still can't find the rule, then consider reporting an issue.
-->

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.

Please consider creating an isolated reproduction repo to make it easy for the volunteer maintainers debug your issue.
-->

```JSON
{
    "plugins": ["@typescript-eslint"],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 2021,
        "sourceType": "module",
        "project": "tsconfig.json",
    },
    "rules": {
        "@typescript-eslint/require-await": "error",
    },
}
```

```json
{
    "compilerOptions": {
        "strict": true
    },
    "include": ["test.ts"]
}
```

```TS
async function* concat<T>(first: Promise<T> | T, it: AsyncIterable<T> | Iterable<T>): AsyncIterableIterator<T> {
    yield first // `yield` unwraps `Promise` (thennable) in async generator function
    yield* it   // `yield*` iterates `AsyncIterable` (having `[Symbol.asyncIterator]()` method) in async generator function
}
```

<!--
Also include your tsconfig, if you're using type-aware linting
-->

**Expected Result**

<!--
What did you expect to happen?
Please be specific here - list the exact lines and messages you expect.
-->

No errors.

**Actual Result**

<!--
What actually happened?
Please be specific here - list the exact lines and messages that caused errors
-->
```
~\dev\sandbox> eslint test.ts

~\dev\sandbox\test.ts
  1:1  error  Async generator function 'concat' has no 'await' expression  @typescript-eslint/require-await

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->
#1691 is related but seems not enough.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.26.1` |
| `@typescript-eslint/parser`        | `4.26.1` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.28.0` |
| `node`                             | `16.1.0` |

--- a/packages/eslint-plugin/src/rules/require-await.ts
+++ b/packages/eslint-plugin/src/rules/require-await.ts
@@ -115,16 +115,22 @@ export default util.createRule({
       if (node?.argument?.type === AST_NODE_TYPES.Literal) {
         // making this `false` as for literals we don't need to check the definition
         // eg : async function* run() { yield* 1 }
-        scopeInfo.isAsyncYield = false;
+        scopeInfo.isAsyncYield ||= false;
       }
 
       const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node?.argument);
       const type = checker.getTypeAtLocation(tsNode);
-      const symbol = type.getSymbol();
-
-      // async function* test1() {yield* asyncGenerator() }
-      if (symbol?.getName() === 'AsyncGenerator') {
-        scopeInfo.isAsyncYield = true;
+      const typesToCheck = expandUnionOrIntersectionType(type);
+      for (const type of typesToCheck) {
+        const asyncIterator = tsutils.getWellKnownSymbolPropertyOfType(
+          type,
+          'asyncIterator',
+          checker,
+        );
+        if (asyncIterator !== undefined) {
+          scopeInfo.isAsyncYield = true;
+          break;
+        }
       }
     }
 
@@ -229,4 +235,11 @@ function getFunctionHeadLoc(
     start,
     end,
   };
+}
+
+function expandUnionOrIntersectionType(type: ts.Type): ts.Type[] {
+  if (type.isUnionOrIntersection()) {
+    return type.types.flatMap(expandUnionOrIntersectionType);
+  }
+  return [type];
 }

--- a/packages/eslint-plugin/tests/rules/require-await.test.ts
+++ b/packages/eslint-plugin/tests/rules/require-await.test.ts
@@ -161,6 +161,57 @@ async function* test1() {
   yield* asyncGenerator();
 }
     `,
+    `
+async function* asyncGenerator() {
+  await Promise.resolve();
+  yield 1;
+}
+async function* test1() {
+  yield* asyncGenerator();
+  yield* 2;
+}
+    `,
+    `
+async function* test(source: AsyncIterable<any>) {
+  yield* source;
+}
+    `,
+    `
+async function* test(source: Iterable<any> & AsyncIterable<any>) {
+  yield* source;
+}
+    `,
+    `
+async function* test(source: Iterable<any> | AsyncIterable<any>) {
+  yield* source;
+}
+    `,
+    `
+type MyType = {
+  [Symbol.iterator](): Iterator<any>;
+  [Symbol.asyncIterator](): AsyncIterator<any>;
+};
+async function* test(source: MyType) {
+  yield* source;
+}
+    `,
+    `
+type MyType = {
+  [Symbol.asyncIterator]: () => AsyncIterator<any>;
+};
+async function* test(source: MyType) {
+  yield* source;
+}
+    `,
+    `
+type MyFunctionType = () => AsyncIterator<any>;
+type MyType = {
+  [Symbol.asyncIterator]: MyFunctionType;
+};
+async function* test(source: MyType) {
+  yield* source;
+}
+    `,
     'const foo: () => void = async function* () {};',
     `
 async function* foo(): Promise<string> {
@@ -281,6 +332,41 @@ const foo = async function* () {
       code: `
 async function* asyncGenerator() {
   yield 1;
+}
+      `,
+      errors: [
+        {
+          messageId: 'missingAwait',
+          data: {
+            name: "Async generator function 'asyncGenerator'",
+          },
+        },
+      ],
+    },
+    {
+      code: `
+async function* asyncGenerator(source: Iterable<any>) {
+  yield* source;
+}
+      `,
+      errors: [
+        {
+          messageId: 'missingAwait',
+          data: {
+            name: "Async generator function 'asyncGenerator'",
+          },
+        },
+      ],
+    },
+    {
+      code: `
+function isAsyncIterable(value: unknown): value is AsyncIterable<any> {
+  return true;
+}
+async function* asyncGenerator(source: Iterable<any> | AsyncIterable<any>) {
+  if (!isAsyncIterable(source)) {
+    yield* source;
+  }
 }
       `,
       errors: [

